src/fsharp/Interactive.DependencyManager/DependencyManager.fs

```diff
@@ -0,0 +1,246 @@
+
+        /// Resolve the dependencies for the given arguments
+        ///
+        member this.ResolveDependencies(scriptDir, mainScriptName, scriptName, scriptExt, packageManagerTextLines, tfm): bool * string seq * string seq * string seq =
+
+            // The ResolveDependencies method, has two signatures, the original signaature in the variable resolveDeps and the updated signature resoveDepsEx
+            // The resolve method can return values in two different tuples:
+            //     (bool * string list * string list * string list)
+            //     (bool * string list * string list)
+            //
+            // We use reflection to get the correct method and to determine what we got back.
+            //
+            let method, arguments =
+                if resolveDepsEx.IsSome then
+                    resolveDepsEx, [| box scriptExt; box packageManagerTextLines; box tfm |]
+                elif resolveDeps.IsSome then
+                    resolveDeps, [| box scriptDir; box mainScriptName; box scriptName; box packageManagerTextLines; box tfm |]
+                else
+                    None, [||]
+
+            let succeeded, references, generatedScripts, additionalIncludeFolders =
+                let result =
+                    match method with
+                    | Some m -> m.Invoke(instance, arguments) :?> _
+                    | None -> false, seqEmpty, seqEmpty, seqEmpty
+
+                // Verify the number of arguments returned in the tuple returned by resolvedependencies, it can be:
+                //     4 - (bool * string list * string list * string list)
+                //     3 - (bool * string list * string list)
+                let tupleFields = result |> FSharpValue.GetTupleFields
+                match tupleFields |> Array.length with
+                | 4 -> tupleFields.[0] :?> bool, tupleFields.[1] :?> string seq, tupleFields.[2] :?> string seq, tupleFields.[3] :?> string seq
+                | 3 -> tupleFields.[0] :?> bool, seqEmpty, tupleFields.[1] :?> string list  |> List.toSeq, tupleFields.[2] :?> string list |> List.toSeq
+                | _ -> false, seqEmpty, seqEmpty, seqEmpty
+
+            succeeded, references, generatedScripts, additionalIncludeFolders
+
+
+type DependencyProvider () =
+
+    // Resolution Path = Location of FSharp.Compiler.Private.dll
+    let assemblySearchPaths = lazy (
+        [
+            let assemblyLocation = typeof<IDependencyManagerProvider>.GetTypeInfo().Assembly.Location
+            yield Path.GetDirectoryName assemblyLocation
+            yield AppDomain.CurrentDomain.BaseDirectory
+        ])
+
+    let enumerateDependencyManagerAssemblies compilerTools reportError =
+        getCompilerToolsDesignTimeAssemblyPaths compilerTools
+        |> Seq.append (assemblySearchPaths.Force())
+        |> Seq.collect (fun path ->
+            try
+                if Directory.Exists(path) then Directory.EnumerateFiles(path, dependencyManagerPattern)
```

Comment:
Since we're already handling an exception, is the `if` check here necessary?

Reply:
The if check eliminates the need for the runtime to generate and throw and catch an exception in the normal use case.  Exceptions are very expensive in runtime, and this is in the middle of a probing loop where the majority of directories may not exist.